external/storm-hbase/src/main/java/org/apache/storm/hbase/trident/windowing/HBaseWindowsStore.java

```diff
@@ -0,0 +1,275 @@
+    }
+
+    @Override
+    public Iterable<Object> get(List<String> keys) {
+        List<Get> gets = new ArrayList<>();
+        for (String key : keys) {
+            WindowsStore.Entry.nonNullCheckForKey(key);
+
+            byte[] effectiveKey = effectiveKey(key);
+            gets.add(new Get(effectiveKey));
+        }
+
+        Result[] results = null;
+        try {
+            results = htable().get(gets);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        Kryo kryo = new Kryo();
+        List<Object> values = new ArrayList<>();
+        for (int i=0; i<results.length; i++) {
+            Result result = results[i];
+            if(result.isEmpty()) {
+                LOG.error("Got empty result for key [{}]", keys.get(i));
+                throw new RuntimeException("Received empty result for key: "+keys.get(i));
+            }
+            Input input = new Input(result.getValue(family, qualifier));
+            Object resultObject = kryo.readClassAndObject(input);
+            values.add(resultObject);
+        }
+
+        return values;
+    }
+
+    @Override
+    public Iterable<String> getAllKeys() {
+        Scan scan = new Scan();
+        // this filter makes sure to receive only Key or row but not values associated with those rows.
+        scan.setFilter(new FirstKeyOnlyFilter());
+        //scan.setCaching(1000);
+
+        final Iterator<Result> resultIterator;
+        try {
+            resultIterator = htable().getScanner(scan).iterator();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        final Iterator<String> iterator = new Iterator<String>() {
+            @Override
+            public boolean hasNext() {
+                return resultIterator.hasNext();
+            }
+
+            @Override
+            public String next() {
+                Result result = resultIterator.next();
+                String key = null;
+                try {
+                    key = new String(result.getRow(), UTF_8);
+                } catch (UnsupportedEncodingException e) {
+                    throw new RuntimeException(e);
+                }
+                return key;
+            }
+
+            @Override
+            public void remove() {
+                throw new UnsupportedOperationException("remove operation is not supported");
+            }
+        };
+
+        return new Iterable<String>() {
+            @Override
+            public Iterator<String> iterator() {
+                return iterator;
+            }
+        };
+    }
+
+    @Override
+    public void put(String key, Object value) {
+        WindowsStore.Entry.nonNullCheckForKey(key);
+        WindowsStore.Entry.nonNullCheckForValue(value);
+
+        if(value == null) {
+            throw new IllegalArgumentException("Invalid value of null with key: "+key);
+        }
+        Put put = new Put(effectiveKey(key));
+        Kryo kryo = new Kryo();
```

Comment:
This creates a new Kryo and Output instance for every invocation which may not be necessary. Also there is no option to register classes so kryo cannot optimize the serialization/deserialization. Its better if you have a serializer interface and provide a default implementation with ability to register classes. You could reuse `org.apache.storm.state.Serializer` and `org.apache.storm.state.DefaultStateSerializer`